### PR TITLE
mgr/dashboard: Add monitor list

### PR DESCRIPTION
**Pull request by Rubab-Syed**
This PR adds a new page for monitor daemon. It shows following metrics:

1. Cluster ID     
2. Monmap modification time
3. Features
4. Monmap epoch
5. Monitor's name 
6. Rank
7. Public IP Address
8. Inline graph of open sessions

### CODE REVIEW

**Comment by jcsp** [comment]
Thanks -- to get this to a mergeable point, please do the following:

1. Rebase your branch on the latest master (this will involve resolving the merge conflict on module.py).  You first fetch the latest code (use "git fetch origin" if your repo was originally cloned from the main ceph repo, or otherwise add a remote ("git remote add upstream git@github.com:ceph/ceph.git") and then fetch from that ("git fetch upstream").  Then, do a "git rebase upstream/master".  This will complain about a conflict in module.py, which you need to edit, then do a "git add" on that file and finally a "git rebase --continue".
2. Squash the commits into larger logical changes (in this PR that probably just means one commit).  I usually do this with "git rebase -i upstream/master", and then mark the smaller commits as "f" so that they are squashed into the commit before them.
3. Edit the commit messages to match the standard commit message style (see other recent commits in git log, the main thing is to start it with the component and a short description ``mgr/dashboard: add monitor listing" and include a "Signed-off-by: Your Name <your email>" at the end) .  I usually do the edit with another "rebase -i" and mark the commit as "r" so that I can edit the commit message.
4. Finally, re-push your branch (same push command as originally, but add a --force).

It's quite a lot of git to just update a PR, but most of the time we're just using these same few commands, so it quickly becomes familiar.  If you already knew all these git commands then I apologise for being so verbose :-)

There's a longer description of the submission guidelines here: https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst (ignore the "Patch submission via ceph-devel@vger.kernel.org" part as we're using pull requests)

**Comment by jcsp** [comment]
I wonder if @jecluis or @tchaikov would have an opinion about the info displayed here? (see bottom of monitors.html for list of columns)

File: src/pybind/mgr/dashboard/monitors.html [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    </h1>
+</section>
+	
+<!-- Main content -->
+<section class="content text-center">
+    <div class="box">
+        <table class="table">
+            <tr style="font-size:15px">
+	         <td><p class="text-left"><span style="color:yellow; font-weight:bold">Cluster ID: </span> {mon_status.monmap.fsid}</p></td>
+	         <td><p class="text-left"><span style="color:yellow; font-weight:bold">Monmap Created: </span> {mon_status.monmap.created}</p></td>
```
[the remark this answers is not in the document]
Comment (jecluis) [reply src/pybind/mgr/dashboard/monitors.html]: Also, I would personally not care much about the creation time of the monmap. I'd much rather only see when it was last modified, as it should also reflect the creation time for the initial monmap.

File: src/pybind/mgr/dashboard/monitors.html [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+	         <td><p class="text-left"><span style="color:yellow; font-weight:bold">Monmap Modified: </span> {mon_status.monmap.modified}</p></td>
+	     </tr>
+        </table>	
+    </div>
+    <div class="box">
+        <h3 style="color:#00bb00; font-weight:bold">INSIDE QUORUM</h3>
```
Comment (jecluis) [inline src/pybind/mgr/dashboard/monitors.html]: I would prefer "IN QUORUM" rather than "INSIDE QUORUM". `Inside` doesn't make much sense in the context of a quorum - you're either in it or you are not.

File: src/pybind/mgr/dashboard/monitors.html [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	 <div class="box-body">
+	     <table class="table table-bordered text-center">
+	         <thead>
+	             <th>Name</th>
+	  	     <th>Rank</th>
+		     <th>IP Addr</th>
+		     <th>Public Addr</th>
```
Comment (jecluis) [inline src/pybind/mgr/dashboard/monitors.html]: `addr` and `public_addr` should be the same, and I'm pretty sure `addr` is only there for legacy reasons. In `MonMap::dump()`, both fields have the same value, and we should always honor `public_addr` in this case. What is missing though (and would be really useful) is showing the monitor's `public_bind_addr`, but I don't think we expose that on the monmap (it's a runtime thing). It should be exposed via  the daemon's `config get`, but I don't think that's reasonable to expect the dashboard to do, really.

In a nutshell, drop `IP Addr`.

File: src/pybind/mgr/dashboard/monitors.html [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+		     <th>Open Sessions</th>
+		     <th>Trimmed Sessions</th>
+		     <th>Elections Participated In</th>
```
Comment (jecluis) [inline src/pybind/mgr/dashboard/monitors.html]: I don't think these fields are particularly helpful. The only thing that would be interesting to know is whether the monitor has ever been in-quorum (in case it is not currently in quorum).

File: src/pybind/mgr/dashboard/monitors.html [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+		     <td><span class="inlinesparkline" style="font-size:30px">{ mon.stats.num_sessions | sparkline_data }</span></td>
+		     <td><span class="inlinesparkline" style="font-size:30px">{ mon.stats.session_trim | sparkline_data }</span></td>
+		     <td><span class="inlinesparkline" style="font-size:30px">{ mon.stats.num_elections | sparkline_data }</span></td>
+		 </tr>
+             </table>
+	 </div>
+    </div>
+
+    <div class="box">
+        <h3 style="color:#00bb00; font-weight:bold">OUTSIDE QUORUM</h3>
```
Comment (jecluis) [inline src/pybind/mgr/dashboard/monitors.html]: "OUTSIDE QUORUM" -> "NOT IN QUORUM"

File: src/pybind/mgr/dashboard/monitors.html [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+	 <div class="box-body">
+	     <table class="table table-bordered text-center">
+	         <thead>
+		     <th>Name</th>
+		     <th>Rank</th>
+		     <th>IP Addr</th>
+		     <th>Public Addr</th>
+		     <th>Open Sessions</th>
+		     <th>Trimmed Sessions</th>
+		     <th>Elections Participated</th>
```
Comment (jecluis) [inline src/pybind/mgr/dashboard/monitors.html]: Same comments from above apply.

File: src/pybind/mgr/dashboard/monitors.html (hunk 4 above)
Comment (jecluis) [reply src/pybind/mgr/dashboard/monitors.html]: Sorry, should have been cleared: I meant the `Trimmed Sessions` and `Elections Participated In`.

**Comment by Rubab-Syed** [comment]
cephmonitor

File: src/pybind/mgr/dashboard/monitors.html [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+<section class="content text-center">
+    <div class="row">
+        <div class="col-md-4">
+	    <div class="box text-left" style="font-size:13.4px">
+                <table class="table table-bordered">
+		    <tr><td><p><span style="color:yellow; font-weight:bold">Cluster ID: </span></td><td>{mon_status.monmap.fsid}</p></td></tr>
+		    <tr><td><p><span style="color:yellow; font-weight:bold">monmap modified: </span></td><td> {mon_status.monmap.modified}</p></td></tr>
+		    <tr><td><p><span style="color:yellow; font-weight:bold">monmap epoch: </span></td><td> {mon_status.monmap.epoch}</p></td></tr>
+		    <tr><td><p><span style="color:yellow; font-weight:bold">quorum con: </span></td><td> {mon_status.features.quorum_con}</p></td></tr>
+		    <tr><td><p><span style="color:yellow; font-weight:bold">required con: </span></td><td> {mon_status.features.required_con}</p></td></tr>
```
Comment (jecluis) [inline src/pybind/mgr/dashboard/monitors.html]: While I think `quorum_con` and `required_con` features are nice values to know, I think the integer values have little value when exposed by themselves to the user on the dashboard. It would be nice to actually have the features being used instead.

It may just so be that you do not have a good way to obtain those; if that is the case, please file a ticket in the tracker mentioning just that, assign it to me, link this PR in the ticket, and I'll address that (and then we can amend this feature to include those features at a later date).

File: src/pybind/mgr/dashboard/monitors.html [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    </div>
+    <div class="box">
+        <h3 style="color:#00bb00; font-weight:bold">NOT IN QUORUM</h3>
+	<div class="box-body">
+	    <table class="table table-bordered text-center">
+	        <thead>
+		    <th>Name</th>
+		    <th>Rank</th>
+		    <th>Public Addr</th>
+		    <th>Open Sessions</th>
```
Comment (jecluis) [inline src/pybind/mgr/dashboard/monitors.html]: There is a fair chance that this field is actually irrelevant if the monitor is not in quorum. I'd suggest dropping it.

**Comment by jecluis** [comment]
Also, Thank you for the screenshot! That helped immensely! :) 👍

**Comment by liewegas** [comment]
I wouldn't worry about the features for now.. we can either drop it or replace it with the release code name.  We don't have a feature bit mask -> string function *except* for a function that tells you which release it looks like (e.g., "luminous", "firefly", etc.), and that's probably not very helpful at this level (the ceph version of the daemon is probably more interesting).

Love the "open sessions" mini plot!

**Comment by liewegas** [comment]
Hmm it would be nice to show the out-of-quorum nodes' sync status (when they are doing a sync).  It's a bit hard because we don't have a percentage value, though (we have the current key position but we don't know how far that is). Also sync is rare-ish.. but when it *does* happen you really want to know how close it is to finishing.

Another thing that would be nice is a "how long in quorum" time value.  I've been pining for a "pretty time" helper for a while that returns things like "32s", "12m", "3.4h", "7w", etc.  Seeing a time in quorum for each mon would make flappy/unreliable mons visible.

(These are all nice-to-haves that can come later of course!)

**Comment by jecluis** [comment]
@liewegas Now that you mention that, I do have a `how long has mon.foo been out of the quorum` on my to-do list. I guess this is as good time as it will ever be to get that going.

File: src/pybind/mgr/dashboard/monitors.html [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+			    <td>{mon.rank}</td>
+			    <td>{mon.public_addr}</td>
+			    <td><span class="inlinesparkline" style="font-size:30px">{ mon.stats.num_sessions | sparkline_data }</span></td>
+			</tr>
+		    </table>
+		</div>
+	    </div>    
+	</div>
+    </div>
+    <div class="box">
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/monitors.html]: i'd suggest move this div right after the the "IN QUORUM" div, i.e after https://github.com/ceph/ceph/pull/19632/files#diff-4e15ad4bfa3f75fa7ca9156de9eedd25R77. so it can be aligned with the "IN QUORUM" table. looks better this way. what do you think?

**Comment by jcsp** [comment]
@jecluis @tchaikov do you guys mind if I merge this, and leave any other tweaks to a follow up PR?
